File: tests/tools/build_test.py
import json
import os
from os import environ, path

import pytest

from tests.tools import mypytest
from tests.tools.build_test_case import BuildTestCase

EXPECTED_OUTPUT = environ.get("EXPECTED_OUTPUT")


class TestBuild(BuildTestCase):

    @pytest.mark.skipif(EXPECTED_OUTPUT is None or len(EXPECTED_OUTPUT) == 0,
                        reason="no output requested")
    def test_output(self):
        expected_output = EXPECTED_OUTPUT + os.linesep
        code, out, err = self.get_output()
        assert (code, out, err) == (0, expected_output, '')

    def test_files(self):
        expected: str = environ.get("EXPECTED_FILES")
        if expected is None or len(expected) == 0:
            return
        expected_dict = json.loads(expected)
        for dirname in expected_dict:
            prefix = None
            external = False
            if len(dirname) > 0 and dirname[0] == "@":
                prefix = [dirname[1:]]
                external = True
            else:
                prefix = [self.output_base]
                if len(dirname) > 0:
                    prefix.append(dirname)

            for f in expected_dict[dirname]:
                assert f is not None

                should_exist = True
                if f[0] == "!":
                    should_exist = False
                    f = f[1:]

                rpath = "/".join(prefix + [f])
                fpath = self.location(rpath, external)
                assert fpath is not None, f'Missing runfile item for {rpath}'
                message_base = f"\n name: {f}\n runfiles path: {rpath}\n file path: {fpath}"
                if should_exist:
                    assert path.exists(fpath), f"Missing expected file: '{f}'" + message_base
                else:
                    assert not path.exists(fpath), f"Expected not to find file: '{f}'" + message_base


if __name__ == "__main__":
    mypytest.main(__file__)

File: tests/tools/build_test.go
package main

import (
	"encoding/json"
	"fmt"
	"github.com/bazelbuild/rules_go/go/tools/bazel"
	"github.com/samhowes/my_rules_dotnet/tests/tools/lib"
	"os"
	"path"
	"strings"
	"testing"
)

var config = lib.TestConfig{}

func initConfig(t *testing.T, config *lib.TestConfig) {
	config.Once.Do(func() {
		var data interface{}

		jsonText := `%config_json%`
		err := json.Unmarshal([]byte(jsonText), &data)
		if err != nil {
			t.Fatalf("failed to get test config: %v", err)
		}
		config.JsonText = jsonText
		config.Data = data.(map[string]interface{})
		config.ExpectedOutput = `%expected_output%`

		var args []string
		argText := `%args%`
		err = json.Unmarshal([]byte(argText), &args)
		if err != nil {
			t.Fatalf("failed to parse executable args: %v", err)
		}

		config.Args = args

		for k := range config.Data {
			fmt.Println(k)
		}
		config.JsonText = jsonText
		config.Target = "%target%"
	})
}

func TestBuildOutput(t *testing.T) {
	initConfig(t, &config)

	fmt.Printf("target: %s\n", config.Target)
	// go_test starts us in our runfiles_tree (on unix) so we can base our assertions off of the current directory
	for dir, filesA := range config.Data["expectedFiles"].(map[string]interface{}) {
		useRunfiles := false
		//workspace := "my_rules_dotnet"

		if len(dir) > 0 && dir[0] == '@' {
			useRunfiles = true
			parts := strings.SplitN(dir[1:], "/", 2)
			//workspace = parts[0]
			if len(parts) == 2 {
				dir = parts[1]
			} else {
				dir = ""
			}
		}

		fmt.Printf("%s\n", dir)
		files := filesA.([]interface{})
		for _, fA := range files {
			f := fA.(string)

			shouldExist := true
			if f[0] == '!' {
				shouldExist = false
				f = f[1:]
			}

			fullPath := path.Join(dir, f)
			if useRunfiles {
				fullPath, err := bazel.Runfile(path.Join(dir, f))
				if shouldExist && err != nil {
					t.Errorf("Failed to find runfile %s: %v", fullPath, err)
					continue
				}
			}

			fmt.Printf(fullPath + "\n")
			_, err := os.Stat(fullPath)
			if !shouldExist && err == nil {
				t.Errorf("expected file to not exist: \n%s", fullPath)
			} else if shouldExist && err != nil {
				t.Errorf("error finding expected file '%s':\n%v", fullPath, err)
			}
		}
	}
}

func TestExecutableOutput(t *testing.T) {
	initConfig(t, &config)
	if config.ExpectedOutput == "" {
		t.Skip("No output requested")
	}

	lib.CheckExecutableOutput(t, &config)
}
